shared/js/ui/views/mixins/unload-listener.es6.js

```diff
@@ -1,27 +1,41 @@
-const browserUIWrapper = require('./../../base/ui-wrapper.es6.js')
+import browser from 'webextension-polyfill'
 
 module.exports = {
     /**
-     * Collect any 'user-action' messages and pass them along
-     * when we call `browserUIWrapper.popupUnloaded` - this allows the background script
-     * to make a decision about whether to reload the current page.
+     * Passes any 'user-action' messages through a messaging connection to the
+     * extension's background, so that they can be used at the point this window
+     * is closed. This allows the background script to make a decision about
+     * whether to reload the current page.
      *
      * For example: If a user has manually disabled protections in the popup,
-     * we don't immediately reload the page behind - but instead we want to wait until the
-     * popup is closed.
+     * we don't immediately reload the page behind - but instead we want to wait
+     * until the popup is closed.
      *
      * @param store
      */
     registerUnloadListener: function (store) {
         /** @type {string[]} */
         const userActions = []
+        let connection = null
+
+        // Create a messaging connection to the background. If the connection is
+        // broken by the background, the background ServiceWorker was restarted
+        // and the connection must be recreated + any earlier user actions
+        // resent.
+        const reconnect = () => {
+            connection = browser.runtime.connect({ name: 'popup' })
+            connection.onDisconnect.addListener(reconnect)
```

Comment:
Probably not a large issue, but do we need to clean up these listeners attached to closed connections?

Reply:
No I don't think so, since the connection is gone along with any of its listeners.